[JIT] script & logging for extracting IR from logs

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72889

The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

Differential Revision: [D34440189](https://our.internmc.facebook.com/intern/diff/D34440189)

davidberard98 committed:
[WIP][JIT] script & logging for extracting IR from logs

The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications

[ghstack-poisoned]

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,148 @@
+    END = "</GRAPH_EXPORT>"
+    pfx = None
+    current = ""
+    graphs = []
```
nit: would this be easier as :
```
split_strs = filename.split(BEGIN)
graphs = []
for split_str in split_strs:
    end = split_str.find(END)
    assert end != -1
    graphs.append(split_str[:end])
```

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,148 @@
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
+        else:
```
let's assert this is a tensor here otherwise we'll get a confusing error message down the line... also we should just add an int input here

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,148 @@
+    return (func, inputs)
+
+
+def run_test(ir, inputs, *, warmup_runs=20, test_runs=20) -> float:
```
warmup_runs == 20 is probably a bit too much, especially if we're not running gradients. Same with 20 test runs.

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,148 @@
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
+        else:
+            inputs.append(torch._C._representative_tensor(inp.type()))
```
Let's assert that the tensors have requires_grad == 0 for now. If they have requires_grad == 1 then we're not currently exercising the gradient part of the graph

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,148 @@
+    for _ in range(warmup_runs):
+        graph(*inputs)
+
+    start_event = torch.cuda.Event(enable_timing=True)
```
In the future we're going to want to fork the timing based on cpu or cuda similar to https://github.com/pytorch/pytorch/blob/master/benchmarks/fuser/run_benchmarks.py#L208

eellison commented on torch/csrc/jit/codegen/cuda/graph_fuser.cpp:
```diff
@@ -1994,6 +1995,20 @@ bool removeInplaceOperations(const std::shared_ptr<Graph>& graph) {
       graph, [&](Node* node) { return inplace_ops.count(node->kind()) != 0; });
 }
 
+void printFusionGroupsForDebug(std::shared_ptr<Graph>& graph) {
```
No need to re-traverse the graph, I would just add this logic here https://github.com/pytorch/pytorch/blob/a04c88f460d6dc461ae25ef7ddfc2687968d6b58/torch/csrc/jit/codegen/cuda/graph_fuser.cpp#L1543

eellison reviewed (commented):
Looks great 😍  😍  😍 

Couple small comments. It's fine if this is tailored to NVFuser but in the future we're going to want to factor this out for autodiff / nnc debugging as well... fine if that doesn't happen here currently.

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

Differential Revision: [D34440189](https://our.internmc.facebook.com/intern/diff/D34440189)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

Differential Revision: [D34440189](https://our.internmc.facebook.com/intern/diff/D34440189)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

davidberard98 committed:
Update on "[JIT] script & logging for extracting IR from logs"


The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

Differential Revision: [D34440189](https://our.internmc.facebook.com/intern/diff/D34440189)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,165 @@
+    inputs = []
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
```
maybe in future diff we can do a range of values for int / float

eellison reviewed (approved):
😍  😍  😍

eellison commented:
cc @jjsjann123 @wconstab

eellison commented on scripts/jit/log_extract.py:
```diff
@@ -0,0 +1,165 @@
+    return graphs
+
+
+def make_tensor_from_type(inp_type: torch._C.TensorType):
```
Hmmm, actually this doesn't use the stride :/ so i think we need to update
